presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java

```diff
@@ -0,0 +1,173 @@
+ * as long as they are simple constants, or they referenced only once (to avoid
+ * introducing duplicate computation) and the references don't appear within
+ * a TRY block (to avoid changing semantics).
+ */
+public class InlineProjections
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!(node instanceof ProjectNode)) {
+            return Optional.empty();
+        }
+
+        ProjectNode parent = (ProjectNode) node;
+
+        PlanNode source = lookup.resolve(parent.getSource());
+        if (!(source instanceof ProjectNode)) {
+            return Optional.empty();
+        }
+
+        ProjectNode child = (ProjectNode) source;
+
+        Sets.SetView<Symbol> candidates = extractCandidates(parent, child);
+        if (candidates.isEmpty()) {
+            return Optional.empty();
+        }
+
+        // inline the expressions
+        Map<Symbol, Expression> parentAssignments = parent.getAssignments()
+                .entrySet().stream()
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        entry -> ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Object>()
+                        {
+                            @Override
+                            public Expression rewriteSymbolReference(SymbolReference reference, Object context, ExpressionTreeRewriter treeRewriter)
+                            {
+                                Symbol symbol = Symbol.from(reference);
+                                if (candidates.contains(symbol)) {
+                                    return child.getAssignments().get(symbol);
+                                }
+
+                                return reference;
+                            }
+                        }, entry.getValue())
+                ));
+
+        // Synthesize identity assignments for the inputs of expressions that were inlined
+        // to place in the child projection.
+        // If all assignments end up becoming identity assignments, they'll get pruned by
+        // other rules
+        Set<Symbol> inputs = child.getAssignments()
+                .entrySet().stream()
+                .filter(e -> candidates.contains(e.getKey()))
+                .map(Map.Entry::getValue)
+                .flatMap(e -> DependencyExtractor.extractAll(e).stream())
+                .collect(toSet());
+
+        Assignments.Builder childAssignments = Assignments.builder();
+        for (Map.Entry<Symbol, Expression> assignment : child.getAssignments().entrySet()) {
+            if (!candidates.contains(assignment.getKey())) {
+                childAssignments.put(assignment.getKey(), assignment.getValue());
+            }
+        }
+        for (Symbol input : inputs) {
+            childAssignments.put(input, input.toSymbolReference());
```

Comment:
is not there a method `Assignments#putIdentity`? Or even `putIdentities`, if not maybe let`s create it.